synapse/federation/federation_base.py

```diff
@@ -80,7 +80,13 @@ async def _check_sigs_and_hash(
           InvalidEventSignatureError if the signature check failed. Nothing
              will be logged in this case.
         """
-        await _check_sigs_on_pdu(self.keyring, room_version, pdu)
+        try:
+            await _check_sigs_on_pdu(self.keyring, room_version, pdu)
+        except Exception as exc:
```

Comment:
It looks like we can just worry about `InvalidEventSignatureError` here 👍

But all of the downstream places in `_check_sigs_on_pdu` do the same `except Exception as e:` 🤷

Reply:
What we decide here, we should also apply to https://github.com/matrix-org/synapse/pull/13814

Is there a way we can make the errors we care about opt-out instead of opt-in? It seems like it's going to suck to keep some of these error lists up to date whenever the downstream functions change and raise different errors.

Maybe it's better to have an opt-in to handle just what you know but as an example at the following spot, we probably need to look at `FederationError`, `RuntimeError`, `InvalidResponseError` from a quick glance at some of the downstream code and this list seems like it will morph under us since there is so much going on in that block. Is there a good way to get a general sense of all exception types that a code path can raise (doesn't need to be comprehensive, maybe info just from docstrings)?

https://github.com/matrix-org/synapse/blob/6f0c3e669da458e838e7b4b165a13e8a5312d6d0/synapse/handlers/federation_event.py#L916-L920

Reply:
> I don't think RuntimeError, ValueError, AssertionError, KeyError etc should be included, for example. These represent programming errors rather than the remote not acting according to the way we expect.

Fair enough on `RuntimeError` and friends, I saw [this case](https://github.com/matrix-org/synapse/blob/ad4c14e4b0c44d6a8ee42e760d7e1fe1755559a2/synapse/event_auth.py#L171-L176 ) and during my quick scan and seemed relevant (it's not) although if it was, would be a good candidate to make it throw the proper error type 👍

> I would suggest trying some cases of where you want to track a failed pull attempt, perhaps even adding tests that reproduce these cases

Having these tests would be great. We'd want to apply them to `/event/{eventId}`, `/event_auth/{roomId}/{eventId}`, `/state/{roomId}?event_id=$xxx`, `/state_ids/{roomId}?event_id=$xxx` (although related and probably covered by the existing backoff and `NotRetryingDestination` in the next point below). For `/backfill` and `/get_missing_events`, we can't track server/network problems because there is no specific event we're requesting but we can track whether the events in their responses are rejected, etc.

> Should a network problem actually be tracked against the event at all; won't the existing backoff behaviour be sufficient there? ....

In general, if anything goes wrong with trying to pull an event from another server, we should backoff and do it in the background next time (https://github.com/matrix-org/synapse/issues/13623).

The existing backoff underlying all of the federation requests (`matrixfederationclient.py` and `NotRetryingDestination`) seems like it could be sufficient for network related errors. I haven't experienced network problems being the cause of `/messages` being slow so I haven't put as much thought in this area.


> Homeserver not on federation whitelist (? what should we do here?)

Seems like we shouldn't be asking homeservers if they aren't on our whitelist in the first place. I guess we just let `FederationDeniedError` handle this and `continue` to the next destination which accomplishes this though.


---

Thanks for the thoughtful response on this @reivilibre! Really helping me form the right thoughts around all of this!